back-end/convert.py

```diff
@@ -0,0 +1,74 @@
+import logging
+import os
+
+import img2pdf
+from PIL import Image
+
+log = logging.getLogger("convert")
+
+
+def change_extension(filename, extension):
+    file, _ = os.path.splitext(filename)
+    return file + extension
+
+
+def get_filename(filepath):
+    return os.path.basename(filepath)
+
+
+def to_jpg(input):
+    output = change_extension(input, ".jpg")
+    if input == output:
+        return get_filename(output)
+
+    log.info("Converting {} to {}".format(
+             get_filename(input), get_filename(output)))
+
+    fill_color = '#000000'
+    image = Image.open(input)
+    if image.mode in ('RGBA', 'LA'):
+        background = Image.new(image.mode[:-1], image.size, fill_color)
+        background.paste(image, image.split()[-1])
+        image = background
+    try:
+        image.save(output)
+        return get_filename(output)
+    except IOError as e:
+        log.error("cannot convert {}, error: {}".format(
```

Comment:
Are you intentionally not doing a `raise` after this `log.error()`? Unless I'm missing something, these converters will all implicitly return `None` on error, and then the Flask endpoint will attempt to return "I converted it! Filename is: None" instead of yielding HTTP 500 or whatnot.

(Explicitly checking for `None` in the Flask code is another option, but I suspect exceptions are more sensible. On the other hand, then you'll log each error twice, as-is. Maybe just remove this type of generic "could not even" exception handling entirely & let the Flask code deal with it? Which dovetails with another comment elsewhere re: that whole flow.)

Reply:
yep I forgot to `raise` those. Not sure I follow, are you suggesting to remove the `except`s here and have the `main` code handle it? I suppose that would make sense

Reply:
I did that. Bit concerned that it would be hard to see where the error orginiated is all

Reply:
Any unhandled exceptions will always show a traceback, so a real deployment of this with logs & such would be pretty easy to diagnose - it'd show an eg IOError originating somewhere inside the Pillow library, passing through the `image.save` line here.

Reply:
Hmm, I think Flask swallows these tbh. I can get around it by manually logging the traceback. Will do that.